### PR TITLE
CREATE-TODO: Completed, CREATE-TODO-TESTS: Completed

### PR DESCRIPTION
**Pull request by Sparsh-Kumar**
## Description
Expose /todo api endpoint to allow clients to add new TODO items

## Database schema changes
A new model named "Todo" was added.

### Automated test cases added
1. POST to /todo endpoint should create a new todo with "title" specified in the request body, should return 201 response code.
2. POST to /todo endpoint should give 400 response code if an empty string is passed as "title" in the request body.
3. POST to /todo endpoint should give 400 response code if a task with same "title" already exists. 

### Manual test cases run
1. URL: /todo
    Method: POST
    RequestBody: {
        "title":"title of my first task"
    }
    ResponseBody: {
        "id":"618ba7a1535bac222dc8d298",
        "title":"title of my first task"
    }
    ResponseCode: 201

2. URL: /todo
    Method: POST
    RequestBody: {
         "title":""
     }
     ResponseBody: {
          "message": "Validation failed. Please modify the request and try again.",
          "failures": [
              {
                 "field": "title",
                  "message": "Please provide the title of the task"
              }
           ]
     }
     ResponseCode: 400

3. URL: /todo
    Method: POST
    RequestBody: {
         "title":"title of my first task"
    }
    ResponseBody: {
    "message": "Validation failed. Please modify the request and try again.",
    "failures": [
        {
            "field": "title",
            "message": "Please provide a different title for the task to be created."
        }
    ]
}
    ResponseCode: 400

### CODE REVIEW

File: src/models/model-factory.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { BaseModel } from './base-model';
 
 export type ModelFactory<T extends BaseModel> = {
-  getType(): string;
+  getType(): any;
```
Comment (jjalan) [inline src/models/model-factory.ts:4]: Why?

File: src/locales/en.json [hunk 2]
```diff
@@ -11,7 +11,8 @@
     "INVALID_FIRSTNAME": "Please specify first name.",
     "INVALID_LASTNAME": "Please specify last name.",
     "PASSWORD_MISSING": "Please provide a password.",
-    "DUPLICATE_EMAIL": "The email you provided is already taken. Please provide a different one."
+    "DUPLICATE_EMAIL": "The email you provided is already taken. Please provide a different one.",
+    "INVALID_TITLE_VALUE": "The title value cannot be empty."
```
Comment (jjalan) [inline src/locales/en.json:15]: This is a poor example of how error messages.

Error messages are not about telling that they are doing something wrong. It's about guiding them what to do to overcome the issue.

Please Google `how to write better error messages`. Ex: https://xd.adobe.com/ideas/process/information-architecture/error-message-design-ux/

File: test/spec/todo/todo.spec.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  app.initializeMiddlewares();
+  app.initializeControllers();
+  app.initializeErrorHandling();
+  expressApp = app.expressApp;
+});
+
+
+describe ('POST /todo',() => {
+
+    it ('should create a new todo with the title provided in the request body', async () => {
```
Comment (jjalan) [inline test/spec/todo/todo.spec.ts:33]: Every test should consists of three steps - validating tests pre-condition, execute test steps and validating tests post condition. In this case:

Pre condition: You need to ensure there is no TODO in database with title: test title
Execute test steps: What you have is correct i.e. making a POST request.
Post condition: What you have is correct but not complete. You also need to ensure that appropriate TODO item is created database.

**Review by jjalan — CHANGES_REQUESTED** [review]
Use 2 instead of 4 spaces across the project.

File: src/models/model-factory.ts (hunk 1 above)
Comment (Sparsh-Kumar) [reply src/models/model-factory.ts:4]: In getModel () in mongo-store.ts, If I was using this statement "modelFactory.getType() === typeof Todo" and returning the 'todo', then I was facing some issues, and was not getting required todo model in return, but always was getting account model in return, so just to resolve this issue and make it simple I did this.

**Review by roopaljasnani — COMMENTED** [review]
While writing code, we also need to make sure it looks clean, readable and formatted in a standard way. This should be a practice we should not avoid. Here are the points you can work on to ensure the same:

- Files should end with a new line (eol=lf)
- Avoid lingering new lines in between of code, unless you are trying to make up for logical code divisions
- Use 2 lines of space across, instead of 4

Refer [here](https://github.com/jalantechnologies/handbook/blob/main/engineering/onboarding-experienced-hire.md#step-2-setup-ide).

**Pull request merged**